File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* root, int &ans) {
        if(!root) return 0;
        int lft= helper(root->left, ans);
        int rgt= helper(root->right, ans);
        ans=max(ans, root->val+max(0,max(lft,rgt)));
        ans=max(ans, max(lft,0)+root->val+max(0, rgt));
        return root->val + max(0, max(lft, rgt));
    }
    int maxPathSum(TreeNode* root) {
        int ans= -1e9;
        helper(root, ans);
        return ans;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    private int maxPathSumUtil(TreeNode root){
        if(root==null) return 0;
        int lft= Math.max(maxPathSumUtil(root.left),0);
        int rgt= Math.max(maxPathSumUtil(root.right),0);
        int val= root.val;
        ans=Math.max(ans,lft+rgt+val);
        return val+Math.max(lft,rgt);
    }
    public int maxPathSum(TreeNode root) {
        ans= Integer.MIN_VALUE;
        maxPathSumUtil(root);
        return ans;
    }
}